### PR TITLE
Fix getodk/central#937: Bubble up iframe message event

### PR DESCRIPTION
**Pull request by sadiqkhoja**
Part of getodk/central#937

#### What has been done to verify that this works as intended?

Added a test to verify event is raised. Manual verification by embedding a form from dev server to a local custom application. Posted the html code for the custom app below.

#### Why is this the best possible solution? Were any other approaches considered?

NA

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

No

#### Does this change require updates to user documentation? If so, please file an issue [here](https://github.com/getodk/docs/issues/new) and include the link below.

Added to https://github.com/getodk/docs/issues/1896

#### Before submitting this PR, please make sure you have:

- [x] run `npm run test` and `npm run lint` and confirmed all checks still pass OR confirm CircleCI build passes
- [x] verified that any code or assets from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

**Comment by sadiqkhoja** [comment]
Sample html code for embedding enketo in ifram as reference. I run `npx live-server` to host it locally.

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Embedded Enketo</title>
  <style>
    iframe {
      width: calc(100vw - 40px);
      height: calc(100vh - 30px);;
    }
  </style>
</head>
<body>
  <iframe src="https://dev.getodk.cloud/-/single/kxHRlsyRlfNZwEMpVhea4r2dogntTpp?st=OqEhm9ftIaQ8KNRQF99srTpELsPT1OZXADEJLCSJICEsenM4JR6dpZwQU2SN9hVJ&parentWindowOrigin=http://127.0.0.1:8080" frameborder="0"></iframe>
   <!-- <iframe src="https://getodk.org" frameborder="0"></iframe> -->
    <script>
      const iframe = document.querySelector('iframe');
      window.addEventListener('message', (event) => {
        console.log('event received');
        console.log(event);

        let eventData;
        try { eventData = JSON.parse(event.data); } catch {}

        if (event.origin === 'https://dev.getodk.cloud' &&
          eventData?.enketoEvent === 'submissionsuccess') {
            window.location = 'https://example.com';
        }

        
      });
    </script>
</body>
</html>
```

**Review by latin-panda — APPROVED** [review]
Thank you for the script!

It looks good!

File: test/components/enketo-iframe.spec.js [hunk 1]
```diff
@@ -75,4 +76,27 @@ describe('EnketoIframe', () => {
...
+
+    const data = JSON.stringify({ enketoEvent: 'submissionsuccess' });
+    const script = document.createElement('script');
+    script.textContent = `window.parent.postMessage('${data}', "*");`;
+    iframe.element.contentDocument.body.appendChild(script);
+
+    await wait();
+
+    postMessage.calledWith(data, window.location.origin).should.be.true;
+  });
```
Comment (matthew-white) [inline test/components/enketo-iframe.spec.js:101]: In the component, `window.parent.postMessage()` is only called if `window.location !== window.parent.location`. Here we see that `postMessage()` ends being called, but I don't totally understand how the two locations differ in this test. I don't see either location being faked. Are the two locations different by default in testing?

File: test/components/enketo-iframe.spec.js
Comment (matthew-white) [inline test/components/enketo-iframe.spec.js:1]: Would it be bad if we bubbled the `message` event in cases where we don't need to? If so, how about adding a test for the negative case? Something to cover: If the origin differs, or if the location is the same, then we shouldn't bubble.

File: src/components/enketo-iframe.vue [hunk 2]
```diff
@@ -87,13 +87,19 @@ watchEffect(() => {
...
-
-  if (event.origin === window.location.origin &&
-    eventData?.enketoEvent === 'submissionsuccess' &&
-    redirectUrl.value) {
-    router.push((new URL(redirectUrl.value)).pathname);
+  if (event.origin === window.location.origin) {
+    // For the cases where this page is embedded in external iframe, pass the event data to the
+    // parent.
+    if (window.location !== window.parent.location) {
+      window.parent.postMessage(event.data, route.query.parentWindowOrigin);
```
[the remark this answers is not in the document]
Comment (sadiqkhoja) [reply src/components/enketo-iframe.vue:94]: `parentWindowOrigin` should be in the `src` of the parent/3rd party application's iframe.

File: test/components/enketo-iframe.spec.js (hunk 1 above)
Comment (sadiqkhoja) [reply test/components/enketo-iframe.spec.js:101]: Because karma uses iframe to load test application the output of 

`console.log(window.location, window.parent.location);`
 is 
`http://localhost:9876/context.html, http://localhost:9876/?id=57567226`

File: test/components/enketo-iframe.spec.js
Comment (sadiqkhoja) [reply test/components/enketo-iframe.spec.js:1]: Added a negative test https://github.com/getodk/central-frontend/pull/1185

File: src/components/enketo-iframe.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthew-white) [reply src/components/enketo-iframe.vue:94]: I see that we're passing `parentWindowOrigin` to `src`, but won't `route.query.parentWindowOrigin` be `undefined` here? `route` is from `useRoute()`, which reflects the route of the Vue app, not the iframe `src`.

Something related I was thinking about: In the test, there's ``mockRouter(`/?parentWindowOrigin=${window.location.origin}`)``, but normally, I don't think `parentWindowOrigin` will be in the route/URL of the Vue app. Or am I misunderstanding that?

Is it possible to call `window.parent.postMessage(event.data, window.location.origin)` without referencing `route.query`?

[the remark this answers is not in the document]
Comment (matthew-white) [reply src/components/enketo-iframe.vue:94]: OK, I think I understand now. In your example above, all of Frontend ends up in an iframe, and `parentWindowOrigin` is indeed passed to Frontend via `route.query`. :+1:

What if the embedding application forgets to specify `parentWindowOrigin` though? How about we check for `route.query.parentWindowOrigin` in the `if` above? Something like:

```js
const { parentWindowOrigin } = route.query;
// route.query values can be array, so I like checking typeof.
if (window.location !== window.parent.location && typeof parentWindowOrigin === 'string') {
  window.parent.postMessage(event.data, parentWindowOrigin);
}
```

[the remark this answers is not in the document]
Comment (sadiqkhoja) [reply src/components/enketo-iframe.vue:94]: added here: https://github.com/getodk/central-frontend/pull/1189
